### PR TITLE
[teamcity] new check

### PR DESCRIPTION
**Pull request by LeoCavaille**
Rebased, improved a little more the PR to fit the current agent codebase.
Also added a test to test the logic of the check : get only newer builds after the agent start. The responses saved in mocks came from a real TeamCity server at one point though.

Supersedes #1171

### CODE REVIEW

File: tests/test_teamcity.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import unittest
+
+# 3p
+from mock import MagicMock, patch
+
+# project
+from tests.common import load_check
+
+CONFIG = {
+    'init_config': {'server': 'localhost:8111'},
```
Comment (remh) [inline tests/test_teamcity.py]: What if you have multiple servers in one host ? 
It seems to me that should be part of the instance configuration, not the init_config

File: conf.d/teamcity.yaml.example [hunk 2]
```diff
@@ -0,0 +1,30 @@
+init_config:
+  # Specify the server name of your teamcity instance here
+  # Guest authentication must be on if you want the check to be able to get data
+  server: teamcity.mycompany.com
```
Comment (remh) [inline conf.d/teamcity.yaml.example]: My comment about "server" being in init_config actually belongs here.

File: checks.d/teamcity.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        if 'server' not in self.init_config:
+            raise Exception("You must specify a server in teamcity.yaml")
+        self.server = self.init_config['server']
+
+    def _initialize_if_required(self, instance_name, build_configuration):
+        # Already initialized
+        if instance_name in self.last_build_ids:
+            return
+
+        self.log.info("Initializing {0}".format(instance_name))
```
Comment (remh) [inline checks.d/teamcity.py]: We don't want that at the info level

File: checks.d/teamcity.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+            raise
+
+        self.log.debug(
+            "Last build id for instance {0} is {1}."
+            .format(instance_name, last_build_id)
+        )
+        self.last_build_ids[instance_name] = last_build_id
+
+    def _build_and_send_event(self, new_build, instance_name, is_deployment, host, tags):
+        self.log.info("Found new build with id {0}, saving and alerting.".format(new_build["id"]))
```
Comment (remh) [inline checks.d/teamcity.py]: It should probably be in debug

File: checks.d/teamcity.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+            raise Exception("Each instance must have a build configuration")
+
+        host = instance.get('host_affected') or self.hostname
+        tags = instance.get('tags')
+        is_deployment = _is_affirmative(instance.get('is_deployment', False))
+
+        self._initialize_if_required(instance_name, build_configuration)
+
+        # Look for new successful builds
+        new_build_url = "http://{0}/guestAuth/app/rest/builds/?locator=buildType:{1},sinceBuild:id:{2},status:SUCCESS"\
```
Comment (remh) [inline checks.d/teamcity.py]: it's a nitpick but let's declare this string as a constant
```python
BUILD_URL = "http://{0}/guestAuth/app/rest/builds/?locator=buildType:{1},sinceBuild:id:{2},status:SUCCESS"
```
and then you can do:
```python
new_build_url = BUILD_URL.format(self.server, build_configuration, self.last_build_ids[instance_name])
```

**Comment by remh** [comment]
It doesn't look like we are collecting build/deployment failures. Is that a limitation of their API ?

It'd be nice to have service checks on those if it's possible to get them.

Otherwise it looks good!

**Comment by LeoCavaille** [comment]
Merging as is. For improvements, we'll see if there is a demand. If you are reading, feel free to file an issue to add more features!

**Pull request merged**